terraform/node_output.go

```diff
@@ -275,16 +275,21 @@ func (n *NodeApplyableOutput) Execute(ctx EvalContext, op walkOperation) (diags
 		// depends_on expressions here too
 		diags = diags.Append(validateDependsOn(ctx, n.Config.DependsOn))
 
-		// Ensure that non-sensitive outputs don't include sensitive values
+		// For root module outputs in particular, an output value must be
+		// statically declared as sensitive in order to dynamically return
+		// a sensitive result, to help avoid accidental exposure in the state
+		// of a sensitive value that the user doesn't want to include there.
 		_, marks := val.UnmarkDeep()
 		_, hasSensitive := marks["sensitive"]
-		if !n.Config.Sensitive && hasSensitive {
-			diags = diags.Append(&hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Output refers to sensitive values",
-				Detail:   "Expressions used in outputs can only refer to sensitive values if the sensitive attribute is true.",
-				Subject:  n.Config.DeclRange.Ptr(),
-			})
+		if n.Addr.Module.IsRoot() {
+			if !n.Config.Sensitive && hasSensitive {
```

Comment:
Not really a big deal, but I think a simpler (and less expensive) way to achieve `hasSensitive` here might be `val.HasMark("sensitive")`, since it seems like we're just discarding the unmarked value above anyway.

Reply:
Unfortunately this would fail to catch nested sensitive values. For example, it would cause this to happen:

```hcl
output "a" {
  value = {
    b = sensitive("foo")
  }
}
```

```
Changes to Outputs:
  + a = {
      + b = "foo"
    }
```

Whereas with the code as-is, we get:

```
│ Error: Output refers to sensitive values
│
│   on main.tf line 1:
│    1: output "a" {
│
│ Expressions used in outputs can only refer to sensitive values if the sensitive attribute is true.
```

Reply:
Ahh yes, right... there isn't a `HasMarkDeep` (or such) method, so we're using `UnmarkDeep` here to get the effect of one. Seems reasonable!